### PR TITLE
Added reports system

### PR DESCRIPTION
**Pull request by navvyua**

### CODE REVIEW

File: app/controllers/admin/reports_controller.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class Admin::ReportsController < AdminController
+  def index
+    @reports = reports
+  end
+
+  def destroy
+    report = Report.find params[:id]
+    report.destroy
```
Comment (IlyaUmanets) [inline app/controllers/admin/reports_controller.rb]: `destroy!`

Comment (IlyaUmanets) [inline app/controllers/admin/reports_controller.rb]: `destroy!`

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -1,33 +1,35 @@
 class CommentsController < ApplicationController
-  before_action :current_resource_user
-  before_action :current_resource_owner
-  before_action :current_resource, only: :destroy
+  before_action :current_commentable
 
   def create
-    @request.comments.create(comment_params)
-    redirect_to [@user, @request]
+    @commentable.comments.create(comment_params)
```
Comment (IlyaUmanets) [inline app/controllers/comments_controller.rb]: `create!`

File: app/controllers/reports_controller.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
+class ReportsController < ApplicationController
+  before_action :current_report_user
+  before_action :current_report, only: [:show, :destroy]
+
+  def index
+    @created_reports = @user.reports.includes(:user).decorate
```
Comment (IlyaUmanets) [inline app/controllers/reports_controller.rb]: create 2 methods (maybe helpers) and move that code there

File: app/controllers/reports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    if @report.save && character
+      redirect_to [@user, @report], notice: t('.report_created')
+    else
+      flash.now[:alert] = t('.incorrect_character')
+      render :new
+    end
+  end
+
+  def destroy
+    @report.destroy
```
Comment (IlyaUmanets) [inline app/controllers/reports_controller.rb]: `destroy!`

File: app/helpers/report_helper.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+module ReportHelper
+  def reports_count
+    Report.all.count
```
Comment (IlyaUmanets) [inline app/helpers/report_helper.rb]: `Report.count`

File: app/models/character.rb [hunk 6]
```diff
@@ -1,10 +1,45 @@
 class Character < ApplicationRecord
   belongs_to :user
 
+  has_many :reports
```
Comment (IlyaUmanets) [inline app/models/character.rb]: `dependent: :delete_all`

File: app/models/character.rb [hunk 7]
```diff
@@ -1,10 +1,45 @@
...
+
   validates_presence_of :first_name, :last_name, :age, :biography
   validates :age, numericality: { only_integer: true, greater_than: 10, less_than: 100 }
```
Comment (IlyaUmanets) [inline app/models/character.rb]: Long line

File: app/models/character.rb [hunk 8]
```diff
@@ -1,10 +1,45 @@
...
+  }
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+
+  class << self
+    private
+
+    def find_by_full_name(name_parts)
```
Comment (IlyaUmanets) [inline app/models/character.rb]: Should be scope

File: app/models/character.rb [hunk 9]
```diff
@@ -1,10 +1,45 @@
...
+      where(
+        'LOWER(characters.first_name) LIKE ? AND LOWER(characters.last_name) LIKE ?',
+        "%#{name_parts[0]}%", "%#{name_parts[1]}%"
+      ).order(:first_name)
+    end
+
+    def find_by_one_part_of_name(name)
```
Comment (IlyaUmanets) [inline app/models/character.rb]: Same

File: app/models/character.rb [hunk 10]
```diff
@@ -1,10 +1,45 @@
...
+      where(
```
Comment (IlyaUmanets) [inline app/models/character.rb]:
```ruby
where(
       'LOWER(characters.first_name) LIKE :q OR LOWER(characters.last_name) LIKE :q',
      q: "%#{name}%"
      ).order(:first_name)
```

File: app/models/report.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+class Report < ApplicationRecord
+  belongs_to :user
+  belongs_to :character
```
Comment (IlyaUmanets) [inline app/models/report.rb]: Sort associations by alphabet

File: app/models/user.rb [hunk 12]
```diff
@@ -7,6 +7,8 @@ class User < ApplicationRecord
   has_many :posts
   has_many :processing_requests, class_name: 'Request', foreign_key: 'admin_id',
                                  dependent: :destroy
+  has_many :reports
```
Comment (IlyaUmanets) [inline app/models/user.rb]: Same for sorting and dependent

File: app/views/admin/reports/_report.html.slim [hunk 13]
```diff
@@ -0,0 +1,12 @@
+tr
+  td = report.id
+  td = link_to report.user.name, profile_user_path(report.user)
+  td = report.description.truncate(70)
+  td.text-center
+    = link_to t('.show'), [report.user, report],
+              class: 'btn btn-flat btn-primary btn-xs'
+    - if policy([:admin, Report]).destroy?
+      |
```
Comment (IlyaUmanets) [inline app/views/admin/reports/_report.html.slim]: Why do we need that `|`

File: app/controllers/admin/reports_controller.rb (hunk 1 above)
Comment (navvyua) [inline app/controllers/admin/reports_controller.rb]: Fixed.

File: app/controllers/comments_controller.rb (hunk 2 above)
Comment (navvyua) [inline app/controllers/comments_controller.rb]: Fixed.

File: app/controllers/reports_controller.rb (hunk 3 above)
Comment (navvyua) [inline app/controllers/reports_controller.rb]: Fixed.

File: app/controllers/reports_controller.rb (hunk 4 above)
Comment (navvyua) [inline app/controllers/reports_controller.rb]: Fixed.

File: app/helpers/report_helper.rb (hunk 5 above)
Comment (navvyua) [inline app/helpers/report_helper.rb]: Fixed, also in other counters.

File: app/models/character.rb (hunk 6 above)
Comment (navvyua) [inline app/models/character.rb]: Added dependent destroy, because reports have comments.

File: app/models/character.rb (hunk 7 above)
Comment (navvyua) [inline app/models/character.rb]: Fixed.

File: app/models/character.rb (hunk 8 above)
Comment (navvyua) [inline app/models/character.rb]: Changed to scope.

File: app/models/character.rb (hunk 9 above)
Comment (navvyua) [inline app/models/character.rb]: Changed to scope.

File: app/models/character.rb (hunk 10 above)
Comment (navvyua) [inline app/models/character.rb]: FIxed.

File: app/models/report.rb (hunk 11 above)
Comment (navvyua) [inline app/models/report.rb]: Fixed.

File: app/models/user.rb (hunk 12 above)
Comment (navvyua) [inline app/models/user.rb]: Can you explain please?

File: app/views/admin/reports/_report.html.slim (hunk 13 above)
Comment (navvyua) [inline app/views/admin/reports/_report.html.slim]: Removed.

File: app/models/user.rb (hunk 12 above)
Comment (IlyaUmanets) [inline app/models/user.rb]: Sort associations by alphabet and add dependent delete

**Pull request merged**